File: app/src/main/java/com/sjn/stamp/ui/fragment/SettingFragment.java
package com.sjn.stamp.ui.fragment;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.preference.EditTextPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.widget.Toast;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.sjn.stamp.R;
import com.sjn.stamp.controller.UserSettingController;
import com.sjn.stamp.ui.DialogFacade;
import com.sjn.stamp.utils.AnalyticsHelper;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.RealmHelper;

public class SettingFragment extends PreferenceFragmentCompat {

    private static final String TAG = LogHelper.makeLogTag(SettingFragment.class);
    private static final int REQUEST_BACKUP = 1;

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        findPreference("import_backup").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                DialogFacade.createConfirmDialog(getActivity(), R.string.dialog_confirm_import, new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        switch (which) {
                            case NEGATIVE:
                                return;
                            case POSITIVE:
                                AnalyticsHelper.trackSetting(getContext(), "import_backup");
                                Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                                intent.addCategory(Intent.CATEGORY_OPENABLE);
                                intent.setType("*/*");
                                startActivityForResult(intent, REQUEST_BACKUP);
                        }
                    }
                }, new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                    }
                }).show();
                return true;
            }
        });

        findPreference("export_backup").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                DialogFacade.createConfirmDialog(getActivity(), R.string.dialog_confirm_export, new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        switch (which) {
                            case NEGATIVE:
                                return;
                            case POSITIVE:
                                AnalyticsHelper.trackSetting(getContext(), "export_backup");
                                ProgressDialog progressDialog = new ProgressDialog(getActivity());
                                progressDialog.setMessage(getString(R.string.message_processing));
                                progressDialog.show();
                                RealmHelper.exportBackUp(getActivity());
                                progressDialog.dismiss();
                        }
                    }
                }, new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                    }
                }).show();
                return true;
            }
        });

        findPreference("licence").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                AnalyticsHelper.trackSetting(getContext(), "licence");
                DialogFacade.createLicenceDialog(getActivity()).show();
                return true;
            }
        });

        findPreference("setting_songs_new_song_days").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                if (preference instanceof EditTextPreference) {
                    ((EditTextPreference) preference).setText(String.valueOf(new UserSettingController().getNewSongDays()));
                }
                return true;
            }
        });
        findPreference("setting_songs_new_song_days").setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                AnalyticsHelper.trackSetting(getContext(), "setting_songs_new_song_days", newValue.toString());
                try {
                    int newSongDays = Integer.parseInt(newValue.toString());
                    if (newSongDays >= 0 && newSongDays <= 999) {
                        new UserSettingController().setNewSongDays(newSongDays);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return false;
            }
        });

        findPreference("setting_songs_most_played_song_size").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                if (preference instanceof EditTextPreference) {
                    ((EditTextPreference) preference).setText(String.valueOf(new UserSettingController().getMostPlayedSongSize()));
                }
                return true;
            }
        });
        findPreference("setting_songs_most_played_song_size").setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                AnalyticsHelper.trackSetting(getContext(), "setting_songs_most_played_song_size", newValue.toString());
                try {
                    int mostPlayedSongSize = Integer.parseInt(newValue.toString());
                    if (mostPlayedSongSize >= 0 && mostPlayedSongSize <= 999) {
                        new UserSettingController().setMostPlayedSongSize(mostPlayedSongSize);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return false;
            }
        });
    }

    @Override
    public void onStart() {
        super.onStart();
        FloatingActionButton mFab = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        if (mFab != null) {
            mFab.setVisibility(View.GONE);
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        FloatingActionButton mFab = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        if (mFab != null) {
            mFab.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        LogHelper.d(TAG, "onActivityResult: requestCode=" + requestCode + ", resultCode=" + resultCode);

        if (requestCode == REQUEST_BACKUP) {
            if (resultCode == AppCompatActivity.RESULT_OK) {
                if (data != null && data.getData() != null && data.getData().getPath().endsWith(".realm")) {
                    ProgressDialog progressDialog = new ProgressDialog(getActivity());
                    progressDialog.setMessage(getString(R.string.message_processing));
                    progressDialog.show();
                    RealmHelper.importBackUp(getActivity(), data.getData());
                    progressDialog.dismiss();
                    DialogFacade.createRestartDialog(getActivity(), new MaterialDialog.SingleButtonCallback() {
                        @Override
                        public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                            getActivity().recreate();
                        }
                    }).show();
                } else {
                    Toast.makeText(getContext(), R.string.invalid_backup_selected, Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

}

File: app/src/main/java/com/sjn/stamp/ui/fragment/SettingFragment.kt
package com.sjn.stamp.ui.fragment

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.view.ViewCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.preference.EditTextPreference
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceFragmentCompat
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.widget.Toast
import com.sjn.stamp.R
import com.sjn.stamp.controller.SongController
import com.sjn.stamp.controller.UserSettingController
import com.sjn.stamp.ui.DialogFacade
import com.sjn.stamp.ui.DrawerMenu
import com.sjn.stamp.ui.activity.MusicPlayerListActivity
import com.sjn.stamp.ui.activity.MusicPlayerListActivity.Companion.START_FRAGMENT_KEY
import com.sjn.stamp.utils.AnalyticsHelper
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.MediaRetrieveHelper
import com.sjn.stamp.utils.RealmHelper
import org.polaric.colorful.Colorful

class SettingFragment : PreferenceFragmentCompat() {

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu?.clear()
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        findPreference("primary_theme")?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
            reboot()
            true
        }
        findPreference("accent_theme")?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
            reboot()
            true
        }
        findPreference("dark_theme")?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            if (newValue is Boolean) {
                Colorful.config(context)
                        .dark(newValue)
                        .apply()
                reboot()
            }
            true
        }
        findPreference("translucent")?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            if (newValue is Boolean) {
                Colorful.config(context)
                        .translucent(newValue)
                        .apply()
                reboot()
            }
            true
        }

        findPreference("song_db_refresh")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            activity?.let {
                DialogFacade.createConfirmDialog(it, R.string.dialog_confirm_song_db_refresh, { _, _ ->
                    context?.let {
                        AnalyticsHelper.trackSetting(it, "song_db_refresh")
                        Thread(Runnable { SongController(it).refreshAllSongs(MediaRetrieveHelper.allMediaMetadataCompat(it, null)) }).start()
                    }
                }).show()
            }
            true
        }

        findPreference("song_db_unknown")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            fragmentManager?.let {
                val transaction = it.beginTransaction()
                transaction.add(R.id.container, UnknownSongFragment())
                transaction.addToBackStack(null)
                transaction.commit()
            }
            true
        }

        findPreference("import_backup")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            activity?.let {
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "*/*"
                startActivityForResult(intent, REQUEST_BACKUP)
            }
            true
        }

        findPreference("export_backup")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            activity?.let {
                DialogFacade.createConfirmDialog(it, R.string.dialog_confirm_export, { _, _ ->
                    context?.let {
                        AnalyticsHelper.trackSetting(it, "export_backup")
                    }
                    ProgressDialog(activity).run {
                        setMessage(getString(R.string.message_processing))
                        show()
                        activity?.let { RealmHelper.exportBackUp(it) }
                        dismiss()
                    }
                }).show()
            }
            true
        }

        findPreference("licence").onPreferenceClickListener = Preference.OnPreferenceClickListener {
            context?.let {
                AnalyticsHelper.trackSetting(it, "licence")
            }
            activity?.let {
                DialogFacade.createLicenceDialog(it).show()
            }
            true
        }

        findPreference("setting_songs_new_song_days").onPreferenceClickListener = Preference.OnPreferenceClickListener { preference ->
            if (preference is EditTextPreference) {
                preference.text = UserSettingController().newSongDays.toString()
            }
            true
        }
        findPreference("setting_songs_new_song_days").onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            context?.let {
                AnalyticsHelper.trackSetting(it, "setting_songs_new_song_days", newValue.toString())
            }
            try {
                val newSongDays = Integer.parseInt(newValue.toString())
                if (newSongDays in 0..999) {
                    UserSettingController().newSongDays = newSongDays
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            false
        }

        findPreference("setting_songs_most_played_song_size").onPreferenceClickListener = Preference.OnPreferenceClickListener { preference ->
            if (preference is EditTextPreference) {
                preference.text = UserSettingController().mostPlayedSongSize.toString()
            }
            true
        }
        findPreference("setting_songs_most_played_song_size").onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            context?.let {
                AnalyticsHelper.trackSetting(it, "setting_songs_most_played_song_size", newValue.toString())
            }
            try {
                val mostPlayedSongSize = Integer.parseInt(newValue.toString())
                if (mostPlayedSongSize in 0..999) {
                    UserSettingController().mostPlayedSongSize = mostPlayedSongSize
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            false
        }
    }

    override fun onStart() {
        super.onStart()
        activity?.let {
            (it.findViewById<View>(R.id.fab) as FloatingActionButton).let {
                ViewCompat.animate(it)
                        .scaleX(0f).scaleY(0f)
                        .alpha(0f).setDuration(100)
                        .start()
            }
        }
    }

    override fun onStop() {
        super.onStop()
        activity?.let {
            (it.findViewById<View>(R.id.fab) as FloatingActionButton).visibility = View.VISIBLE
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        LogHelper.d(TAG, "onActivityResult: requestCode=$requestCode, resultCode=$resultCode")

        if (requestCode == REQUEST_BACKUP) {
            if (resultCode == AppCompatActivity.RESULT_OK) {
                activity?.let { activity ->
                    if (data?.data?.path?.endsWith(".realm") == true) {
                        DialogFacade.createConfirmDialog(activity, R.string.dialog_confirm_import, { _, _ ->
                            context?.let {
                                AnalyticsHelper.trackSetting(it, "import_backup")
                                ProgressDialog(it).run {
                                    setMessage(getString(R.string.message_processing))
                                    show()
                                    RealmHelper.importBackUp(activity, data.data)
                                    dismiss()
                                }
                                DialogFacade.createRestartDialog(it) { _, _ ->
                                    reboot()
                                }.show()
                            }
                        }).show()
                    } else {
                        Toast.makeText(activity, R.string.invalid_backup_selected, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    private fun reboot() {
        activity?.finish()
        startActivity(Intent(activity, MusicPlayerListActivity::class.java).apply { putExtra(START_FRAGMENT_KEY, DrawerMenu.SETTING.menuId) })
    }

    companion object {

        private val TAG = LogHelper.makeLogTag(SettingFragment::class.java)
        private const val REQUEST_BACKUP = 1
    }

}
